### DISCUSSION TITLE
Automerge is not working with config copied from docs

### DISCUSSION THREAD

**Discussion by McPringle** [renovatebot/renovate#13273] [Request Help] [answered]
This is my `renovate.json`:

```json
{
  "extends": [
    "config:base",
    ":disableDependencyDashboard"
  ],
  "ignorePaths": [
    "Dockerfile",
    "package.json",
    "pnpm-lock.yaml"
  ],
  "assignees": [
    "McPringle"
  ],
  "reviewers": [
    "McPringle"
  ],
  "packageRules": [
    {
      "matchUpdateTypes": ["minor", "patch", "pin", "digest"],
      "automerge": true
    },
    {
      "matchDepTypes": ["devDependencies"],
      "automerge": true
    }
  ]
}
```

I copied the `packageRules` from the renovate docs about automerge. Now I have a patch update for my `pom.xml`, which should be merged automatically (change: 2.6.1 -> 2.6.2). But the PR stays open. In the description generated by the renovate bot I found a message that automerge is enabled:

```
🚦 Automerge: Enabled.
```

My repo does not require a review. Regarding the GitHub message next to the `Merge pull request` button, the PR can be merged without any problem.

```
This branch has no conflicts with the base branch
Merging can be performed automatically.
```

Here is a link to the PR, which I kept open in case you want to take a look at it:
https://github.com/komunumo/komunumo-server/pull/178

I have absolutely no clue, why the renovate bot does not want to merge the PR automatically. Unfortunately, I could not find helpful information while searching the docs, SO, Google, etc.

**Comment by McPringle** [comment]
I want to add a few lines of the renovate log:

```
DEBUG: Checking if we can automerge branch(branch="renovate/ch.qos.logback-logback-classic-1.x")
DEBUG: mergeStatus=no automerge(branch="renovate/ch.qos.logback-logback-classic-1.x")
```

And a few lines later:

```
DEBUG: PR is configured for automerge(branch="renovate/ch.qos.logback-logback-classic-1.x")
```

I have no idea, why the merge status is `no automerge`...

**Comment by viceice** [comment]
Renovate isn't allowed to merge the pr, probaly blocked by branch protection

**Reply by McPringle** [reply]
Thank you very much for your answer. I don't have branch protection rules:

```
No branch protection rules defined yet.
```

But searching all available settings pages I found these two options:
![image](https://user-images.githubusercontent.com/1254039/147281999-544eb8e6-5e2b-4976-8652-c21f91656ea1.png)
Both were disabled (looks like this is the default setting). I enabled them and now I'm waiting for the next PR coming in to validate this.

Maybe it would be a good idea to mention these options in the renovate doc. As I verified, these are disabled by default.

**Reply by rarkins** [reply]
Please screenshot the full branch protection page

**Reply by McPringle** [reply]
Last night the next PR comes in, it was a minor update. Automerge did not work. Here is a screenshot from the PR:

![image](https://user-images.githubusercontent.com/1254039/147403823-8bd69772-4cca-4575-a608-b1bfdf0939d3.png)

As you can see, the update is minor and automerge is enabled in the configuration. GitHub says the merge can be done automatically. Automerge setting in the GitHub repository is enabled, too:

![image](https://user-images.githubusercontent.com/1254039/147403843-c3626a1c-a2a4-4704-ab76-b60a93e01ea5.png)

In this repository, there are (and have never been) any branch protection rules:

![image](https://user-images.githubusercontent.com/1254039/147403858-33a6e950-7e06-4fbf-97a1-933aedfd5ae2.png)

The log of the renovate bot says the PR is not ready for merge but does not log the reason:

![image](https://user-images.githubusercontent.com/1254039/147404061-478f0d1c-bcf7-4850-8c0d-19f924cc231e.png)

I have absolutely no idea, why this is not working. Every setting I check is okay and the doc lacks information about what settings are needed for the repository. Any ideas?

**Reply by rarkins** [reply]
Something I'd recommend is setting platformAutomerge=true in your config. That way when PRs are created, GitHub's automerge will be activated whenever a PR is configured for automerge, like the one in your screenshot. This will result in faster merging in all situations

**Reply by McPringle** [reply]
But that will automatically merge *all* PRs, right? I don't want to merge major updates as there are often breaking changes and compatibility issues. In my project I use Vaadin and major releases often have a kind of a migration guide.

**Reply by rarkins** [reply]
No, it will only enable platform automerge if the PR is already automerging.

**Reply by McPringle** [reply]
Okay, I'll try that.

**Comment by McPringle** [comment]
Still not merging! Today another PR was created. You can see it [here](https://github.com/komunumo/komunumo-server/pull/208). The change is *not* major:

```
- <version>3.0.0-M1</version>
+ <version>3.0.0-M2</version>
```

The PR itself says, that auto-merge is enabled:

```
🚦 Automerge: Enabled.
```

Regarding GitHub, the merging can be performed automatically:

![image](https://user-images.githubusercontent.com/1254039/147824729-44dbff84-13ff-4841-84f6-ee9dc2983e96.png)

No branch protection rules (as you can see in a previous comment), GitHub merge config is okay, but renovate still does not automatically merge. The renovate log contains these two interesting lines:

```
DEBUG: Checking if we can automerge branch(branch="renovate/org.apache.maven.plugins-maven-deploy-plugin-3.x")
DEBUG: mergeStatus=no automerge(branch="renovate/org.apache.maven.plugins-maven-deploy-plugin-3.x")
```

No more information about why it does not automatically merge.

```json
{
  "extends": [
    "config:base",
    ":disableDependencyDashboard"
  ],
  "ignorePaths": [
    "Dockerfile",
    "package.json",
    "pnpm-lock.yaml"
  ],
  "assignees": [
    "McPringle"
  ],
  "reviewers": [
    "McPringle"
  ],
  "packageRules": [
    {
      "matchUpdateTypes": ["minor", "patch", "pin", "digest"],
      "automerge": true
    },
    {
      "matchDepTypes": ["devDependencies"],
      "automerge": true
    }
  ],
  "platformAutomerge": true
}
```

Has anybody got the auto-merge feature working?

**Reply by viceice** [reply]
Platform automerge only works with required status checks.

You need to allow automerge in repo settings too.

**Reply by McPringle** [reply]
Everything I could find in the settings is enabled:
![image](https://user-images.githubusercontent.com/1254039/147825077-bcc40b9f-0796-4229-b870-7780a71e57b1.png)

I have no branch protection rules:
![image](https://user-images.githubusercontent.com/1254039/147825166-7a52a2e6-9421-4acb-821e-279970d0c77d.png)

So you suggest adding branch protection rules? Okay, I'll try it. Does any doc exist which explains how to configure auto-merge successfully? Actually, it feels like poking around until at some time I hopefully get a config working by accident.

**Reply by viceice** [reply]
https://docs.renovatebot.com/configuration-options/#platformautomerge

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request

**Reply by McPringle** [reply]
Checking out the branch protection rules and status checks, the GitHub docs forward me to a REST API. So I have to set up a service running anywhere on a server that calls the GitHub API to do a status check to enable renovate to automatically merge? Holy cow! Do I understand this correctly?

**Reply by viceice** [reply]
You don't need to touch the API, renovate will check it. But for GitHub automerge you need branch protection rules.

**Reply by McPringle** [reply]
Thanks for the links, but the renovate docs does not show what settings have to be configured to successfully use auto-merge. I have read the docs about `platformAutomerge` and `packageRules`. I tried everything stated in the docs. But I did not find a line about which settings have to be enabled or disabled on GitHub. Just:

> `platformAutomerge` will configure PRs to be merged after all (if any) branch policies have been met. This option is available for Azure, GitHub and GitLab. It falls back to Renovate-based automerge if the platform-native automerge is not available.

I don't have any branch policies. So it should work: *to be merged after all (if any) branch policies have been met*

The GitHub doc about "Automatically merging a pull request": As my screenshots above show, auto-merge is enabled and allowed.

**Reply by McPringle** [reply]
Okay, thanks @viceice, I created a rule and now I'm waiting for the next PR coming in:
![image](https://user-images.githubusercontent.com/1254039/147825929-6379cc90-30b3-4b80-8e8a-cfce07f1379f.png)

Thank you very much for your help, I cross my fingers... :-)

Oh, and a **Happy New Year**!

**Reply by viceice** [reply]
You can close the existing PR, delete the branch and rename the PR. Then renovate should open a new PR .😉

**Reply by rarkins** [reply]
Do you have any tests? The last screenshot you posted implies that you do not. Neither Renovate or GitHub's automerge capabilities work by default unless you have at least one passing test.

**Reply by McPringle** [reply]
@rarkins I have 22 unit tests in my project and all are passing.
@viceice I activated "Require status checks to pass before merging" and "Require branches to be up to date before merging" for my `main` branch and now renovate successfully merged automatically!

Thank you very much for all your help and a Happy New Year!

**Reply by rarkins** [reply]
Do you know why GitHub said it couldn't find any status checks in the last week in the repo?

**Reply by McPringle** [reply]
Absolutely no idea.

**Reply by rarkins** [reply]
There's a good chance it's related

**Comment by McPringle** [comment, marked as answer]
In this answer I collect all information that got the auto-merge feature working:

1. In `renovate.json`, configure `packageRules` with `automerge` and activate `platformAutomerge`:
```json
{
  "extends": [
    "config:base"
  ],
  "packageRules": [
    {
      "matchUpdateTypes": ["minor", "patch", "pin", "digest"],
      "automerge": true
    },
    {
      "matchDepTypes": ["devDependencies"],
      "automerge": true
    }
  ],
  "platformAutomerge": true
}
```

2. In the GitHub repository settings, go to `Option` and activate at least `Allow auto-merge` and `Automatically delete head branches`:
![image](https://user-images.githubusercontent.com/1254039/147834858-341bc165-7569-4374-806e-16e83368d25a.png)

3. In the GitHub repository settings, go to `Branches` and add a branch protection rule for your `main` or `master` branch (whatever you are using). Activate `Require status checks to pass before merging` and `Require branches to be up to date before merging`. If GitHub tells you "No status checks found", just ignore that. Rules must be saved explicitly (scroll down to the "Save" button).
![image](https://user-images.githubusercontent.com/1254039/147835058-35f48191-6ef9-4633-b064-5abdae1f210c.png)

The next PR from the renovate bot should be merged automatically.

**Reply by simao-silva** [reply]
Is this configuration still valid? Suddenly mine stop working using both Renovate from Github and self-hosted. Now it always opens a PR and automerge is disabled.
My `renovate.json` that was working: 
```
{
  "extends": [
    "config:base",
    ":disableDependencyDashboard"
  ],
  "packageRules": [
    {
      "matchUpdateTypes": ["major", "minor", "patch"],
      "automerge": true,
      "automergeType": "branch"
    }
  ],
  "platformAutomerge": true
}
```

**Reply by McPringle** [reply]
The configuration I documented above is still working on my GitHub repos. You configured `"automergeType": "branch"`, I use the default (which is `pr`).

**Reply by rarkins** [reply]
It's valid. Start a new discussion with relevant logs if you want to discuss your situation.
